### PR TITLE
Res 1350 weight decay experiments

### PR DESCRIPTION
**Pull request by lucasosouza**
Experiments with weight decay:
- Adds prefix name to resnet layers following pytorch convention, easier to iterate using named_parameters
-  Reimplements batch norm for weight decay based on `named_parameters`. Implements batch norm for bias using a similar approach.

Let me know if there any issues with adding this extra prefix to the name. Other functions could be simplified using `named_parameters`, such as `init_resnet50_batch_norm`.

Fun and random fact: removing weight decay from bias is exactly the use case that motivated the [creation of named_parameters](https://github.com/pytorch/pytorch/pull/1242) in the first place. Please read the docs/issue/PR for details on named_parameters.

Extra:
- Updates to dynamic_sparse/browser.py, helper method used to navigate results in a notebook (old code not part of core functions or library; review appreciated but not required)

### CODE REVIEW

File: nupic/research/frameworks/pytorch/imagenet/experiment_utils.py [hunk 1]
```diff
@@ -212,33 +210,25 @@ def create_optimizer(model, optimizer_class, optimizer_args,
...
-            if is_bn:
-                group_args.update(weight_decay=0.0)
-
-            model_params.append(group_args)
+
+    group_decay, group_no_decay = [], []
+    for name, param in model.named_parameters():
+        if not param.requires_grad:
+            continue
+        if ".bn" in name and not batch_norm_weight_decay:
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py:224]: That is a very specific pattern, ".bn" (notice the dot). We are also using a very specific network, which is under our control, there are no surprises here. I don't see why we should complicate what is simple.

[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py:224]: There is an issue here, I can't check module type. I only get name and the tensor using named parameters. It is either this or change to modules as it was before.

[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py:224]: If it is specific to the network then add a method to the network class and call that method instead of this function. From the specific network class you can freely access any of the network modules by name or directly.

The current version of this function is not specific to the network implementation, you can pass any ResNet implementation (pytorch or custom) and it will update the batchnorm modules using the module type without any problems, you can adapt the iterator logic to also update the bias.

Relying on naming convention can cause many unintended problems and it is really hard to maintain.

[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py:224]: Thanks for the snippet, that is helpful. Just to be clear, I am not introducing a dependency on naming - there are references elsewhere on that same file, such as `m.bn2.weight` or `m.regular_path`, which is why I didn't think it would go against the conventions to use naming in that function as well. In general, I agree with being generalizable as long as we are consistent throughout the codebase.

[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py:224]: In the examples you mentioned the code is checking the object instance type before accessing a member property just like you would do with any other python class.
Which is different than selecting parameters from a list of parameters assuming the parameter name matching partial string.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py:224]: For those non-generalizable cases such as `n.bn2.weight`, notice they're in a function called `init_resnet50_batch_norm` and are proceeded by checks such as `isinstance(m, torchvision.models.resnet.BasicBlock)`. Thus, they are actually serving a relatively specific purpose. One could argue that `create_optimizer` is intended to be more general.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 2]
```diff
@@ -229,11 +229,13 @@ def setup_experiment(self, config):
         optimizer_class = config.get("optimizer_class", torch.optim.SGD)
         optimizer_args = config.get("optimizer_args", {})
         batch_norm_weight_decay = config.get("batch_norm_weight_decay", True)
+        bias_weight_decay = config.get("bias_weight_decay", True)
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py:225]: Added

File: nupic/research/frameworks/dynamic_sparse/common/browser.py [hunk 3]
```diff
@@ -164,17 +216,19 @@ def _read_experiment(experiment_state, experiment_path):
             with open(params_file) as f:
                 params[exp_tag] = json.load(f)
 
+    # import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/dynamic_sparse/common/browser.py]: Removed

File: nupic/research/frameworks/pytorch/imagenet/experiment_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/experiment_utils.py]: Implemented, following suggestion.
